src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1080,10 +1081,35 @@ RouterInsertJob(Query *originalQuery, Query *query, DeferredErrorMessage **plann
 	Job *job = NULL;
 	bool requiresMasterEvaluation = false;
 	bool deferredPruning = false;
+	bool isMultiRowInsert = false;
 
-	if (!CanShardPrune(distributedTableId, query))
+	RangeTblEntry *valuesRTE = ExtractDistributedInsertValuesRTE(originalQuery);
+	if (valuesRTE != NULL)
 	{
-		/* there is a non-constant in the partition column, cannot prune yet */
+		/*
+		 * We expand the values_lists to contain all default expressions
+		 * from the target list. By doing this early on, in the original
+		 * query, we can later evaluate default expressions for each
+		 * individual row and then perform shard pruning.
+		 */
+		valuesRTE->values_lists = ExpandValuesLists(originalQuery->targetList,
+													valuesRTE->values_lists);
+
+		isMultiRowInsert = true;
+	}
+
+	if (isMultiRowInsert || !CanShardPrune(distributedTableId, query))
```

Comment:
This bookkeeping boolean is more confusing than simply inlining the `valuesRTE != NULL` in this `if` statement. Can you switch to that?

Reply:
Also, couldn't `CanShardPrune` just incorporate this check? Because—after this PR—wouldn't it be true that multi-row `INSERT`s cannot be shard-pruned?